File: lib/cli/commands/Claim.ts
import {
  extractRefundPublicKeyFromReverseSwapTree,
  extractRefundPublicKeyFromSwapTree,
} from 'boltz-core';
import { Arguments } from 'yargs';
import { constructClaimTransaction } from '../../Core';
import { getHexBuffer, stringify } from '../../Utils';
import BuilderComponents from '../BuilderComponents';
import { prepareTx } from '../Command';

export const command =
  'claim <network> <preimage> <privateKey> <redeemScript> <rawTransaction> <destinationAddress> [feePerVbyte] [blindingKey] [discountCT]';

export const describe = 'claims reverse submarine or chain to chain swaps';

export const builder = {
  network: BuilderComponents.network,
  preimage: BuilderComponents.preimage,
  privateKey: BuilderComponents.privateKey,
  redeemScript: BuilderComponents.redeemScript,
  rawTransaction: BuilderComponents.rawTransaction,
  destinationAddress: BuilderComponents.destinationAddress,
  feePerVbyte: BuilderComponents.feePerVbyte,
  blindingKey: BuilderComponents.blindingKey,
  discountCT: BuilderComponents.discountCT,
};

export const handler = async (argv: Arguments<any>): Promise<void> => {
  const {
    keys,
    walletStub,
    swapOutput,
    transaction,
    redeemScript,
    destinationAddress,
  } = await prepareTx(argv, [
    extractRefundPublicKeyFromSwapTree,
    extractRefundPublicKeyFromReverseSwapTree,
  ]);

  const claimTransaction = constructClaimTransaction(
    walletStub,
    [
      {
        ...swapOutput,
        keys,
        redeemScript,
        txHash: transaction.getHash(),
        preimage: getHexBuffer(argv.preimage),
      } as any,
    ],
    destinationAddress,
    argv.feePerVbyte,
  ).toHex();

  console.log(stringify({ claimTransaction }));
};

File: boltzr-cli/src/tx/claim.rs
use crate::tx::utils::construct_transaction;
use anyhow::Result;
use boltz_core::{Network, Transaction, utils::InputType};

#[allow(clippy::too_many_arguments)]
pub fn claim_utxo(
    network: Network,
    preimage: [u8; 32],
    private_key: [u8; 32],
    swap_tree_or_redeem_script: &str,
    raw_transaction: Vec<u8>,
    destination_address: &str,
    fee_per_vbyte: f64,
    blinding_key: Option<[u8; 32]>,
) -> Result<Transaction> {
    construct_transaction(
        network,
        InputType::Claim(preimage),
        private_key,
        swap_tree_or_redeem_script,
        raw_transaction,
        destination_address,
        fee_per_vbyte,
        blinding_key,
    )
}
